File: Zoom/ZMCreateaMeeting/ZMCreateaMeeting.vb
Imports Ayehu.Sdk.ActivityCreation.Interfaces
Imports Ayehu.Sdk.ActivityCreation.Extension
Imports System.Data
Imports System.Net
Imports System.Net.Http
Imports Newtonsoft.Json
Imports Newtonsoft.Json.Linq
Imports System.Text
Imports System
Imports System.Collections.Generic
Imports System.Web

Namespace Ayehu.Sdk.ActivityCreation
    Public Class CustomActivity
        Implements IActivityAsync

        Public ApiKey as string
Public ApiSecret as string

        Public userId as string
Public topic as string
Public type as string
Public start_time as string
Public duration as string
Public timezone as string
Public password as string
Public agenda as string
Public field as string
Public value as string
Public repeat_interval as string
Public weekly_days as string
Public monthly_day as string
Public monthly_week as string
Public monthly_week_day as string
Public end_times as string
Public end_date_time as string
Public host_video as string
Public participant_video as string
Public cn_meeting as string
Public in_meeting as string
Public join_before_host as string
Public mute_upon_entry as string
Public watermark as string
Public use_pmi as string
Public approval_type as string
Public registration_type as string
Public audio as string
Public auto_recording as string
Public enforce_login as string
Public enforce_login_domains as string
Public alternative_hosts as string
Public close_registration as string
Public waiting_room as string
Public items as string
Public contact_name as string
Public contact_email as string
Public registrants_email_notification as string
Public meeting_authentication as string
Public authentication_option as string
Public authentication_domains as string


        Dim Httpmethod As String = "post"
        Dim client As HttpClient = New HttpClient()
        Dim dt As DataTable = New DataTable("resultSet")
        Dim RequestURL As String
        Dim TableName As String = "meetings"
        Public Async Function Execute() As System.Threading.Tasks.Task(Of ICustomActivityResult) Implements IActivityAsync.Execute

            ServicePointManager.Expect100Continue = True
            ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12
            ServicePointManager.ServerCertificateValidationCallback = New System.Net.Security.RemoteCertificateValidationCallback(AddressOf AcceptAllCertifications)
            Dim PostData As String = "{  ""topic"" : " & Microsoft.VisualBasic.Strings.Chr(34) & topic & Microsoft.VisualBasic.Strings.Chr(34) & ",  ""type""  : " & type & ",  ""start_time"" : " & Microsoft.VisualBasic.Strings.Chr(34) & start_time & Microsoft.VisualBasic.Strings.Chr(34) & ",  ""duration"" : " & Microsoft.VisualBasic.Strings.Chr(34) & duration & Microsoft.VisualBasic.Strings.Chr(34) & ",  ""timezone"" : " & Microsoft.VisualBasic.Strings.Chr(34) & timezone & Microsoft.VisualBasic.Strings.Chr(34) & ",  ""password"" : " & Microsoft.VisualBasic.Strings.Chr(34) & password & Microsoft.VisualBasic.Strings.Chr(34) & ",  ""agenda"" : " & Microsoft.VisualBasic.Strings.Chr(34) & agenda & Microsoft.VisualBasic.Strings.Chr(34) & ",  ""recurrence"": {    ""type""  : " & type & ",    ""repeat_interval"" : " & Microsoft.VisualBasic.Strings.Chr(34) & repeat_interval & Microsoft.VisualBasic.Strings.Chr(34) & ",    ""weekly_days"" : " & Microsoft.VisualBasic.Strings.Chr(34) & weekly_days & Microsoft.VisualBasic.Strings.Chr(34) & ",    ""monthly_day"" : " & Microsoft.VisualBasic.Strings.Chr(34) & monthly_day & Microsoft.VisualBasic.Strings.Chr(34) & ",    ""monthly_week"" : " & Microsoft.VisualBasic.Strings.Chr(34) & monthly_week & Microsoft.VisualBasic.Strings.Chr(34) & ",    ""monthly_week_day"" : " & Microsoft.VisualBasic.Strings.Chr(34) & monthly_week_day & Microsoft.VisualBasic.Strings.Chr(34) & ",    ""end_times"" : " & Microsoft.VisualBasic.Strings.Chr(34) & end_times & Microsoft.VisualBasic.Strings.Chr(34) & ",    ""end_date_time"" : " & Microsoft.VisualBasic.Strings.Chr(34) & end_date_time & Microsoft.VisualBasic.Strings.Chr(34) & "  },  ""settings"": {    ""host_video"" : " & Microsoft.VisualBasic.Strings.Chr(34) & host_video & Microsoft.VisualBasic.Strings.Chr(34) & ",    ""participant_video"" : " & Microsoft.VisualBasic.Strings.Chr(34) & participant_video & Microsoft.VisualBasic.Strings.Chr(34) & ",    ""cn_meeting"" : " & Microsoft.VisualBasic.Strings.Chr(34) & cn_meeting & Microsoft.VisualBasic.Strings.Chr(34) & ",    ""in_meeting"" : " & Microsoft.VisualBasic.Strings.Chr(34) & in_meeting & Microsoft.VisualBasic.Strings.Chr(34) & ",    ""join_before_host"" : " & Microsoft.VisualBasic.Strings.Chr(34) & join_before_host & Microsoft.VisualBasic.Strings.Chr(34) & ",    ""mute_upon_entry"" : " & Microsoft.VisualBasic.Strings.Chr(34) & mute_upon_entry & Microsoft.VisualBasic.Strings.Chr(34) & ",    ""watermark"" : " & Microsoft.VisualBasic.Strings.Chr(34) & watermark & Microsoft.VisualBasic.Strings.Chr(34) & ",    ""use_pmi"" : " & Microsoft.VisualBasic.Strings.Chr(34) & use_pmi & Microsoft.VisualBasic.Strings.Chr(34) & ",    ""approval_type"" : " & Microsoft.VisualBasic.Strings.Chr(34) & approval_type & Microsoft.VisualBasic.Strings.Chr(34) & ",    ""registration_type"" : " & Microsoft.VisualBasic.Strings.Chr(34) & registration_type & Microsoft.VisualBasic.Strings.Chr(34) & ",    ""audio"" : " & Microsoft.VisualBasic.Strings.Chr(34) & audio & Microsoft.VisualBasic.Strings.Chr(34) & ",    ""auto_recording"" : " & Microsoft.VisualBasic.Strings.Chr(34) & auto_recording & Microsoft.VisualBasic.Strings.Chr(34) & ",    ""enforce_login"" : " & Microsoft.VisualBasic.Strings.Chr(34) & enforce_login & Microsoft.VisualBasic.Strings.Chr(34) & ",    ""enforce_login_domains"" : " & Microsoft.VisualBasic.Strings.Chr(34) & enforce_login_domains & Microsoft.VisualBasic.Strings.Chr(34) & ",    ""alternative_hosts"" : " & Microsoft.VisualBasic.Strings.Chr(34) & alternative_hosts & Microsoft.VisualBasic.Strings.Chr(34) & ",    ""global_dial_in_countries"": [      """"    ],    ""registrants_email_notification"" : " & Microsoft.VisualBasic.Strings.Chr(34) & registrants_email_notification & Microsoft.VisualBasic.Strings.Chr(34) & "  }}"
            Dim Data = New StringContent(PostData, Encoding.UTF8, "application/json")
			Dim queryString As Object = HttpUtility.ParseQueryString(String.Empty)
            RequestURL="https://api.zoom.us/v2/users/" & userId & "/meetings?"& queryString.ToString()

                                

            If RequestURL.EndsWith("&") OrElse RequestURL.EndsWith("?") Then RequestURL = RequestURL.Substring(0, RequestURL.Length - 1)
            Dim content As System.Net.Http.StringContent = New StringContent(PostData, Encoding.UTF8, "application/json")
            content.Headers.ContentType = New System.Net.Http.Headers.MediaTypeHeaderValue("application/json")
            client.DefaultRequestHeaders.Add("authorization", "Bearer " + JsonWebToken.Encode(ApiKey, ApiSecret))

            Dim Request As HttpResponseMessage = Nothing
            Select Case Httpmethod
                Case "get"
                    Request = client.GetAsync(RequestURL).Result
                Case "delete"
                    Request = client.DeleteAsync(RequestURL).Result
                Case "put"
                    Request = client.PutAsync(RequestURL, Data).Result
                Case "post"
                    Request = client.PostAsync(RequestURL, Data).Result
                Case "patch"
                    Dim myHttpRequestMessage As New HttpRequestMessage(New HttpMethod("PATCH"), RequestURL)
                    myHttpRequestMessage.Content = content
                    Request = client.SendAsync(myHttpRequestMessage).Result
            End Select
            Select Case Request.StatusCode

                Case HttpStatusCode.NoContent, HttpStatusCode.Created, HttpStatusCode.Accepted
                    Return GenerateActivityResult("Success")

                Case HttpStatusCode.OK

                    Dim values As JObject = JsonConvert.DeserializeObject(Of JObject)(Request.Content.ReadAsStringAsync().Result)
                    If String.IsNullOrEmpty(TableName) Then  'Single record
                        For Each obj As Object In values
                            If dt.Columns.Count = 0 Then
                                dt.Columns.Add(obj.Key)
                            Else
                                Exit For
                            End If
                        Next
                        Dim newrow As DataRow = dt.NewRow
                        For Each obj As Object In values
                            If dt.Columns.Contains(obj.Key) = False Then dt.Columns.Add(obj.Key) ' Some columns discovered later
                            newrow(obj.Key) = obj.Value
                        Next
                        dt.Rows.Add(newrow)
                    Else
                        If values.ContainsKey(TableName) Then
                            For Each usersJ As JObject In values(TableName)
                                For Each obj As Object In usersJ
                                    If dt.Columns.Count = 0 Then
                                        dt.Columns.Add(obj.Key)
                                    Else
                                        Exit For
                                    End If
                                Next

                                Dim newrow As DataRow = dt.NewRow
                                For Each obj As Object In usersJ
                                    If dt.Columns.Contains(obj.Key) = False Then dt.Columns.Add(obj.Key) ' Some columns discovered later
                                    newrow(obj.Key) = obj.Value
                                Next
                                dt.Rows.Add(newrow)
                            Next
                        Else
                            Dim newrow As DataRow = dt.NewRow
                            For Each obj As Object In values
                                If dt.Columns.Contains(obj.Key) = False Then dt.Columns.Add(obj.Key) ' Some columns discovered later
                                newrow(obj.Key) = obj.Value
                            Next
                            dt.Rows.Add(newrow)
                        End If
                    End If

                    Return GenerateActivityResult(dt)

                Case Else
                    If String.IsNullOrEmpty(Request.ReasonPhrase) = False Then
                        Throw New System.Exception(Request.ReasonPhrase)
                    Else
                        Throw New System.Exception(Request.StatusCode)
                    End If

            End Select

        End Function

        Public Function AcceptAllCertifications(ByVal sender As Object, ByVal certification As System.Security.Cryptography.X509Certificates.X509Certificate, ByVal chain As System.Security.Cryptography.X509Certificates.X509Chain, ByVal sslPolicyErrors As System.Net.Security.SslPolicyErrors) As Boolean
            Return True
        End Function

    End Class

    Public Class JsonWebToken

        Public Shared Function Encode(APIKey As String, Secret As String) As String

            Dim keyBytes As Byte() = System.Text.Encoding.UTF8.GetBytes(Secret)

            Dim myhdr As New HDR
            myhdr.alg = "HS256"
            myhdr.typ = "JWT"

            Dim Expiry As System.DateTime = DateTime.UtcNow.AddSeconds(120)
            Dim ts As Integer = CInt((Expiry - New System.DateTime(1970, 1, 1)).TotalSeconds)
            Dim myhdrpayload As String = Newtonsoft.Json.JsonConvert.SerializeObject(myhdr)
            Dim mytpayload As New Tpayload
            mytpayload.iss = APIKey
            mytpayload.exp = ts

            Dim mypayload As String = Newtonsoft.Json.JsonConvert.SerializeObject(mytpayload)

            Dim headerBytes As Byte() = System.Text.Encoding.UTF8.GetBytes(myhdrpayload)
            Dim payloadBytes As Byte() = System.Text.Encoding.UTF8.GetBytes(mypayload)

            Dim stringToSign = Base64UrlEncode(headerBytes) + "." + Base64UrlEncode(payloadBytes)
            Dim bytesToSign = System.Text.Encoding.UTF8.GetBytes(stringToSign)
            Dim sha As New System.Security.Cryptography.HMACSHA256(keyBytes)
            Dim signature As Byte() = sha.ComputeHash(bytesToSign)

            Return stringToSign + "." + Base64UrlEncode(signature)

        End Function

        Private Shared Function Base64UrlEncode(ByVal input As Byte()) As String
            Dim output = System.Convert.ToBase64String(input)
            output = output.Split("="c)(0)
            output = output.Replace("+"c, "-"c)
            output = output.Replace("/"c, "_"c)
            Return output
        End Function
    End Class

    Public Class HDR

        Public alg As String
        Public typ As String

    End Class

    Public Class Tpayload

        Public iss As String
        Public exp As Integer

    End Class

End Namespace

File: Zoom/Meetings/ZM Create a Meeting/ZM Create a Meeting.cs
using System;
using Ayehu.Sdk.ActivityCreation.Interfaces;
using Ayehu.Sdk.ActivityCreation.Extension;
using Ayehu.Sdk.ActivityCreation.Helpers;
using System.Net;
using System.Net.Http;
using System.Text;
using System.Collections.Generic;

namespace Ayehu.Zoom
{
    public class ZM_Create_a_Meeting : IActivityAsync
    {


    
    public string Jsonkeypath = "meetings";
    
    public string apikey = "";
    
    public string password1 = "";
    
    public string userId = "";
    
    public string topic = "";
    
    public string type_p = "";
    
    public string start_time = "";
    
    public string duration = "";
    
    public string timezone = "";
    
    public string password = "";
    
    public string agenda = "";
    
    public string tracking_fields = "";
    
    public string recurrence_type = "";
    
    public string repeat_interval = "";
    
    public string weekly_days = "";
    
    public string monthly_day = "";
    
    public string monthly_week = "";
    
    public string monthly_week_day = "";
    
    public string end_times = "";
    
    public string end_date_time = "";
    
    public string host_video = "";
    
    public string participant_video = "";
    
    public string cn_meeting = "";
    
    public string in_meeting = "";
    
    public string join_before_host = "";
    
    public string mute_upon_entry = "";
    
    public string watermark = "";
    
    public string use_pmi = "";
    
    public string approval_type = "";
    
    public string registration_type = "";
    
    public string audio = "";
    
    public string auto_recording = "";
    
    public string enforce_login = "";
    
    public string enforce_login_domains = "";
    
    public string alternative_hosts = "";
    
    public string close_registration = "";
    
    public string waiting_room = "";
    
    public string contact_name = "";
    
    public string contact_email = "";
    
    public string registrants_email_notification = "";
    
    public string meeting_authentication = "";
    
    public string authentication_option = "";
    
    public string authentication_domains = "";
    
    private bool omitJsonEmptyorNull = true;
    
    private string contentType = "application/json";
    
    private string endPoint = "https://api.zoom.us";
    
    private string httpMethod = "POST";
    
    private string _uriBuilderPath;
    
    private string _postData;
    
    private System.Collections.Generic.Dictionary<string, string> _headers;
    
    private System.Collections.Generic.Dictionary<string, string> _queryStringArray;
    
    private string uriBuilderPath {
        get {
            if (string.IsNullOrEmpty(_uriBuilderPath)) {
_uriBuilderPath = string.Format("v2/users/{0}/meetings",userId);
            }
return _uriBuilderPath;
        }
        set {
            this._uriBuilderPath = value;
        }
    }
    
    private string postData {
        get {
            if (string.IsNullOrEmpty(_postData)) {
_postData = string.Format("{{ \"topic\": \"{0}\",  \"type\": \"{1}\",  \"start_time\": \"{2}\",  \"duration\": \"{3}\",  \"timezone\": \"{4}\",  \"password\": \"{5}\",  \"agenda\": \"{6}\",  \"tracking_fields\": {7},  \"recurrence\": {{   \"type\": \"{8}\",    \"repeat_interval\": \"{9}\",    \"weekly_days\": \"{10}\",    \"monthly_day\": \"{11}\",    \"monthly_week\": \"{12}\",    \"monthly_week_day\": \"{13}\",    \"end_times\": \"{14}\",    \"end_date_time\": \"{15}\"   }},  \"settings\": {{   \"host_video\": \"{16}\",    \"participant_video\": \"{17}\",    \"cn_meeting\": \"{18}\",    \"in_meeting\": \"{19}\",    \"join_before_host\": \"{20}\",    \"mute_upon_entry\": \"{21}\",    \"watermark\": \"{22}\",    \"use_pmi\": \"{23}\",    \"approval_type\": \"{24}\",    \"registration_type\": \"{25}\",    \"audio\": \"{26}\",    \"auto_recording\": \"{27}\",    \"enforce_login\": \"{28}\",    \"enforce_login_domains\": \"{29}\",    \"alternative_hosts\": \"{30}\",    \"close_registration\": \"{31}\",    \"waiting_room\": \"{32}\",    \"contact_name\": \"{33}\",    \"contact_email\": \"{34}\",    \"registrants_email_notification\": \"{35}\",    \"meeting_authentication\": \"{36}\",    \"authentication_option\": \"{37}\",    \"authentication_domains\": \"{38}\"   }} }}",topic,type_p,start_time,duration,timezone,password,agenda,tracking_fields,recurrence_type,repeat_interval,weekly_days,monthly_day,monthly_week,monthly_week_day,end_times,end_date_time,host_video,participant_video,cn_meeting,in_meeting,join_before_host,mute_upon_entry,watermark,use_pmi,approval_type,registration_type,audio,auto_recording,enforce_login,enforce_login_domains,alternative_hosts,close_registration,waiting_room,contact_name,contact_email,registrants_email_notification,meeting_authentication,authentication_option,authentication_domains);
            }
return _postData;
        }
        set {
            this._postData = value;
        }
    }
    
    private System.Collections.Generic.Dictionary<string, string> headers {
        get {
            if (_headers == null) {
_headers = new Dictionary<string, string>() { {"authorization","Bearer " + AyehuHelper.JWTToken(apikey,password1,"HS256","JWT", 120)} };
            }
return _headers;
        }
        set {
            this._headers = value;
        }
    }
    
    private System.Collections.Generic.Dictionary<string, string> queryStringArray {
        get {
            if (_queryStringArray == null) {
_queryStringArray = new Dictionary<string, string>() {  };
            }
return _queryStringArray;
        }
        set {
            this._queryStringArray = value;
        }
    }
    
    public ZM_Create_a_Meeting() {
    }
    
    public ZM_Create_a_Meeting(
                string Jsonkeypath, 
                string apikey, 
                string password1, 
                string userId, 
                string topic, 
                string type_p, 
                string start_time, 
                string duration, 
                string timezone, 
                string password, 
                string agenda, 
                string tracking_fields, 
                string recurrence_type, 
                string repeat_interval, 
                string weekly_days, 
                string monthly_day, 
                string monthly_week, 
                string monthly_week_day, 
                string end_times, 
                string end_date_time, 
                string host_video, 
                string participant_video, 
                string cn_meeting, 
                string in_meeting, 
                string join_before_host, 
                string mute_upon_entry, 
                string watermark, 
                string use_pmi, 
                string approval_type, 
                string registration_type, 
                string audio, 
                string auto_recording, 
                string enforce_login, 
                string enforce_login_domains, 
                string alternative_hosts, 
                string close_registration, 
                string waiting_room, 
                string contact_name, 
                string contact_email, 
                string registrants_email_notification, 
                string meeting_authentication, 
                string authentication_option, 
                string authentication_domains) {
        this.Jsonkeypath = Jsonkeypath;
        this.apikey = apikey;
        this.password1 = password1;
        this.userId = userId;
        this.topic = topic;
        this.type_p = type_p;
        this.start_time = start_time;
        this.duration = duration;
        this.timezone = timezone;
        this.password = password;
        this.agenda = agenda;
        this.tracking_fields = tracking_fields;
        this.recurrence_type = recurrence_type;
        this.repeat_interval = repeat_interval;
        this.weekly_days = weekly_days;
        this.monthly_day = monthly_day;
        this.monthly_week = monthly_week;
        this.monthly_week_day = monthly_week_day;
        this.end_times = end_times;
        this.end_date_time = end_date_time;
        this.host_video = host_video;
        this.participant_video = participant_video;
        this.cn_meeting = cn_meeting;
        this.in_meeting = in_meeting;
        this.join_before_host = join_before_host;
        this.mute_upon_entry = mute_upon_entry;
        this.watermark = watermark;
        this.use_pmi = use_pmi;
        this.approval_type = approval_type;
        this.registration_type = registration_type;
        this.audio = audio;
        this.auto_recording = auto_recording;
        this.enforce_login = enforce_login;
        this.enforce_login_domains = enforce_login_domains;
        this.alternative_hosts = alternative_hosts;
        this.close_registration = close_registration;
        this.waiting_room = waiting_room;
        this.contact_name = contact_name;
        this.contact_email = contact_email;
        this.registrants_email_notification = registrants_email_notification;
        this.meeting_authentication = meeting_authentication;
        this.authentication_option = authentication_option;
        this.authentication_domains = authentication_domains;
    }


        public async System.Threading.Tasks.Task<ICustomActivityResult> Execute()
        {

            HttpClient client = new HttpClient();
            ServicePointManager.Expect100Continue = true;
            ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12;
            ServicePointManager.ServerCertificateValidationCallback = new System.Net.Security.RemoteCertificateValidationCallback(AcceptAllCertifications);
            UriBuilder UriBuilder = new UriBuilder(endPoint); 
            UriBuilder.Path = uriBuilderPath;
            UriBuilder.Query = AyehuHelper.queryStringBuilder(queryStringArray);
            HttpRequestMessage myHttpRequestMessage = new HttpRequestMessage(new HttpMethod(httpMethod), UriBuilder.ToString());
           
            if (contentType == "application/x-www-form-urlencoded")
                myHttpRequestMessage.Content = AyehuHelper.formUrlEncodedContent(postData);
            else
              if (string.IsNullOrEmpty(postData) == false)
                if (omitJsonEmptyorNull)
                    myHttpRequestMessage.Content = new StringContent(AyehuHelper.omitJsonEmptyorNull(postData), Encoding.UTF8, "application/json");
                else
                    myHttpRequestMessage.Content = new StringContent(postData, Encoding.UTF8, contentType);


            foreach (KeyValuePair<string, string> headeritem in headers)
                client.DefaultRequestHeaders.Add(headeritem.Key, headeritem.Value);

            HttpResponseMessage response = client.SendAsync(myHttpRequestMessage).Result;

            switch (response.StatusCode)
            {
                case HttpStatusCode.NoContent:
                case HttpStatusCode.Created:
                case HttpStatusCode.Accepted:
                case HttpStatusCode.OK:
                    {
                        if (string.IsNullOrEmpty(response.Content.ReadAsStringAsync().Result) == false)
                            return this.GenerateActivityResult(response.Content.ReadAsStringAsync().Result, Jsonkeypath);
                        else
                            return this.GenerateActivityResult("Success");
                    }
                default:
                    {
                        if (string.IsNullOrEmpty(response.Content.ReadAsStringAsync().Result) == false)
                            throw new Exception(response.Content.ReadAsStringAsync().Result);
                        else if (string.IsNullOrEmpty(response.ReasonPhrase) == false)
                            throw new Exception(response.ReasonPhrase);
                        else
                            throw new Exception(response.StatusCode.ToString());
                    }
            }
        }

        public bool AcceptAllCertifications(object sender, System.Security.Cryptography.X509Certificates.X509Certificate certification, System.Security.Cryptography.X509Certificates.X509Chain chain, System.Net.Security.SslPolicyErrors sslPolicyErrors)
        {
            return true;
        }
    }
}